File: Ad-Hoc/__1000__/1147/1147.lua
-- coordenadas das cadas adjacentes que o cavalo pode alcançar 
local KNIGHTDELTA = {
    {-2, -1}, {-2,  1},     -- cima esquerda e cima direita
    {-1,  2}, {1,   2},     -- direita cima e direita baixo
    {2,  -1}, {2,   1},     -- baixo esquerda e baixo direita
    {-1, -2}, {1,  -2}      -- esquerda cima e esquerda baixo
}

-- coordenada as casas adjacentes que o peão pode alcança.
local PAWNDELTA = {
    {1, -1},                 -- diagonal esquerda
    {1,  1}                  -- diagonal direita
}

local function knightScape(knightPosition, iteration)
    local dangerousPos = {}
    
    -- lê as coordenadas dos 8 peões 
    for i=1, 8 do
        local pawnpos = io.read()
        local row = 9 - tonumber(string.sub(pawnpos, 1, 1))
        local col = string.byte(pawnpos, 2, 2) - 96 

        for i=1, #PAWNDELTA do 
            local nx = row + PAWNDELTA[i][1]
            local ny = col + PAWNDELTA[i][2]
            if nx >= 1 and nx <= 8 and ny >= 1 and ny <= 8 then
                table.insert(dangerousPos, string.format("%d %d", nx, ny))
            end
        end
    end

    local ans = 0
    local KR = 9 - tonumber(string.sub(knightPosition, 1, 1))
    local KC = string.byte(knightPosition, 2, 2) - 96
    for i=1, #KNIGHTDELTA do
        local nx = KR + KNIGHTDELTA[i][1]
        local ny = KC + KNIGHTDELTA[i][2]
        if nx >= 1 and nx <= 8 and ny >= 1 and ny <= 8 then
            local pos = string.format("%d %d", nx, ny)
            local safePos = true
            for i=1, #dangerousPos do
                if pos == dangerousPos[i] then 
                    safePos = false
                    break
                end
            end
            if safePos then ans = ans + 1 end
        end
    end

    print(string.format("Caso de Teste #%d: %d movimento(s).", iteration, ans))
end

local function main()
    local i = 1
    while true do
        local knightPosition = io.read()
        if knightPosition == "0" then break end

        knightScape(knightPosition, i)
        i = i + 1
    end
end

main() 

File: Ad-Hoc/__1000__/1147/1147.cpp
#include <bits/stdc++.h>
using namespace std;

const int PAWNDELTA[2][2] = { {-1, -1}, {-1, 1} };
const int HORSEDELTA[8][2] = {
    {-2, -1}, {-2,  1}, {-1, -2}, {-1,  2},
    { 1, -2}, { 1,  2}, { 2, -1}, { 2,  1}};

int main() {
    int T = 1;
    string pos;
    while ((cin >> pos) && (pos != "0")) {
        int horse_x = pos[0] - '1';
        int horse_y = pos[1] - 'a';

        set<pair<int, int>> dangerousPos;
        for (size_t i = 0; i < 8; i++) {
            cin >> pos;
            int pawn_x = pos[0] - '1';
            int pawn_y = pos[1] - 'a';
            for (auto &delta : PAWNDELTA) {
                int px = pawn_x + delta[0];
                int py = pawn_y + delta[1];
                if (px < 0 || px > 7 || py < 0 || py > 7) continue;
                dangerousPos.insert({px, py});
            }
        }

        int ans = 0;
        for (auto &delta : HORSEDELTA) {
            int hx = horse_x + delta[0];
            int hy = horse_y + delta[1];
            if (hx < 0 || hx > 7 || hy < 0 || hy > 7) continue;
            if (dangerousPos.count({hx, hy})) continue;
            ans++;
        }

        cout << "Caso de Teste #" << T++ << ": " << ans << " movimento(s).\n";
    }
}